File: facade/src/main/java/org/tdf/sunflower/facade/Validator.java
package org.tdf.sunflower.facade;

import org.tdf.sunflower.types.Block;
import org.tdf.sunflower.types.Transaction;
import org.tdf.sunflower.types.ValidateResult;

public interface Validator extends BlockValidator, PendingTransactionValidator {
    Validator NONE = new Validator() {
        @Override
        public ValidateResult validate(Block block, Block dependency) {
            return ValidateResult.success();
        }

        @Override
        public ValidateResult validate(Block dependency, Transaction transaction) {
            return ValidateResult.success();
        }
    };
}

File: facade/src/main/java/org/tdf/sunflower/facade/Validator.kt
package org.tdf.sunflower.facade

import org.tdf.sunflower.types.Block
import org.tdf.sunflower.types.ValidateResult
import org.tdf.sunflower.types.ValidateResult.Companion.success

/**
 * validate blocks
 */
interface BlockValidator {
    /**
     * validate the block
     *
     * @param block      the block to validate
     * @param dependency parent block
     * @return validate result
     */
    fun validate(rd: RepositoryReader, block: Block, dependency: Block): ValidateResult
}


interface Validator : BlockValidator {
    companion object {
        val NONE: Validator = object : Validator {
            override fun validate(rd: RepositoryReader, block: Block, dependency: Block): ValidateResult {
                return success()
            }
        }
    }
}